Feature/67413 toggle facet visibility

a-rousseau opened the pull request:
[AB#67413](https://dev.azure.com/orckestra001/da965986-c85b-43be-9915-219568b000e5/_workitems/edit/67413)
Add the toggle button
Add functionalities to expand collapse the filters side bar
Add functionalities to change the number of products per line when collapsing or expanding
Adjust the styling for the new functionality
Add new localized text for the toggle button
Fix the navtab styling and HTML markup in Search page
Adjust the facets styling and HTML markup
Apply modifications to Content search as well as Product Search
Added the color swatches in the Product Details page
Added popover with the color name when doing a mouseover on the color swatches 
![Screen Shot 2022-11-17 at 3 34 24 PM](https://user-images.githubusercontent.com/44198946/202554691-6931bd7a-c7e1-4662-986f-3e8951923c04.png)

**Sticky after scrolling**
![Screen Shot 2022-11-17 at 3 34 40 PM](https://user-images.githubusercontent.com/44198946/202554714-5cfe903e-5763-4efe-9b07-b00dd583a0f0.png)

**Hide filters**
![Screen Shot 2022-11-17 at 3 34 56 PM](https://user-images.githubusercontent.com/44198946/202554724-fb2b5e60-d290-432c-84d5-def64b76ea3b.png)

**Product Details Page color swatches**
![Screen Shot 2022-11-17 at 3 34 06 PM](https://user-images.githubusercontent.com/44198946/202554735-2072a59d-33fb-4901-84f7-ce77be2f5c2f.png)

**Color swatches with hover information popover**
![Screen Shot 2022-11-23 at 11 14 27 AM](https://user-images.githubusercontent.com/44198946/203595532-19631ed0-84ce-48b9-859a-3f887be519dd.png)

![Screen Shot 2022-11-23 at 11 17 08 AM](https://user-images.githubusercontent.com/44198946/203596129-c240d301-7772-4998-ae67-d6e7a4003ea0.png)

Commit:
Create a toggle view for facets

AB#67413
Add the toggle button
Add functionalities to expand collapse the filters side bar
Add functionalities to change the number of products per line when collapsing or expanding
Adjust the styling for the new functionality
Add new localized text for the toggle button
Fix the navtab styling and HTML markup in Search page
Adjust the facets styling and HTML markup
Apply modifications to Content search as well as Product Search

Commit:
Merge branch 'dev' into feature/67413-Toggle-Facet-Visibility

Commit:
Make facets and pagination toolbar sticky

Make the facets column sticky and scrollable
Make the pagination toolbar sticky

Commit:
Add ability to save Facet visiblity state

Added a service to allow saving the facets in the user session
Fixed some display issues on mobile
Hid the bottom pagination because we have a sticky one now

Commit:
Fix facets visibility for mobile

Fix facets not showing on mobile size if it is set to Hidden on desktop

Commit:
Fix facets and header overlap on desktop

Fixed a z-index issue when on desktop and scrolling down, the facets were on top of the header

Commit:
Fix KVA Display in product details page

Fix KVA display to be properly aligned and spaced using flexboxes

Commit:
Fix Product Details variants and styling

Fixed the styling for the variants list display
Added the new color selector for the variants
Changed the cursor on hover for the image to a zoom icon

InnaBoitsun commented on Build/install.cake:
```diff
@@ -247,6 +247,19 @@ Task("Create-Https-Binding").Does(() =>
+            .SetIpAddress("*")
+            .SetCertificateHash(certificate.GetCertHash())
+            .SetCertificateStoreName("My")
+    );
```
do we need this changes in Build script?

a-rousseau replied:
Oops ! this is my local build setup, It got added by mistake, will remove this

Commit:
Revert Install script

Revert Install script

Commit:
Update facets styling when browsing categories

Remove bglighter from the classes on the facets card

InnaBoitsun commented on src/Orckestra.Composer.Website/App_Data/Razor/Composer/Product/Summary.cshtml:
```diff
@@ -67,64 +68,114 @@
\ No newline at end of file
+        }
+
+        @helper KvaList()
```
is it used? looks like you have commented    @* @KvaList() *@

a-rousseau replied:
not used

Commit:
Remove unused code

Remove unused code from Product Summary function

InnaBoitsun commented on src/Orckestra.Composer.Website/App_Data/Razor/Composer/Product/Summary.cshtml:
```diff
@@ -41,7 +41,7 @@
             </div>
             <div class="col-md-5 col-lg-4">
-                
+
```
if no changes, better to rollback new lines

InnaBoitsun commented on src/Orckestra.Composer.Website/UI.Package/Sass/Lib/bootstrap/_nav.scss:
```diff
@@ -32,7 +32,7 @@
 .nav-tabs {
-  border-bottom: $nav-tabs-border-width solid $nav-tabs-border-color;
+  //border-bottom: $nav-tabs-border-width solid $nav-tabs-border-color;
 
```
if not needed just remove

a-rousseau replied:
Going to override it as this is a base bootstrap file

InnaBoitsun commented on src/Orckestra.Composer.Website/UI.Package/Templates/KvaItems.hbs:
```diff
@@ -1,58 +1,114 @@
 
             <h5 class="font-weight-bold">{{DisplayName}}</h5>
+            <div class="kva-list variant-kva">
+              {{#if_eq PropertyName "Colour"}}
```
at some point we need to migrate this to razor  :)

a-rousseau replied:
I tried but failed :(

Commit:
Remove whitespace changes

Remove whitespace chanages

Commit:
Update Summary.cshtml

fix spaces

Commit:
Update Summary.cshtml

fix whitespaces

Commit:
Update Summary.cshtml

Fix whitespaces

Commit:
Override Bootstrap lib instead of editing

Moved css definition in an override

Commit:
Fix some bugs

Fix homepage product tiles resizing from 4 to 3 items in a row in the Product Queries component
Move KvaItems to Razor function
Remove HBS KvaItems

Commit:
Fix display of disabled kvas

Made it more apparent that kvas were disabled by adding a line crossing the box diagonally
Really disabled the click on the kva in the click handler
Fixed the expressions in the Vue component bindings
Added some classes to facilitate element identification
Removed unused image tag as we have a separate color tag now to handle image swatches

Commit:
remove tracing

remove console.log tracing

Commit:
add popover on color swatches

Add popover with color name information on all color swatches
Removed commented code in pagination

Commit:
Fix wrong variable call

Fix value.Title to v.Title on kva color swatch

Commit:
Remove double title in the popover

Remove the title because it was displayed twice in the popover

InnaBoitsun commented on src/Orckestra.Composer.Website/UI.Package/Typescript/Composer.ContentSearch/SearchResultsController.ts:
```diff
@@ -33,21 +35,64 @@ module Orckestra.Composer {
+                            document.getElementById("leftCol").classList.remove("w-0-lg");
+                            rightCol.classList.add("col-lg-9");
+                        }
+                        self.vueSearchResults.$data.FacetsVisible = FacetsVisible;
```
it is better have boolean Vue property and here have 
this.FacetsVisible  = !this.FacetsVisible;

Commit:
Fix toggle logic

Made the checked variables Boolean - FacetsVisible and the cookie setShowFacets
Added method hideFacet for clarity
Added method showFacet for clarity

Commit:
Update boolean logic

Update to simplify the logic

Commit:
Final logic fixes

Fix some issues when the view was not updated

Commit:
Remove unused markup

Remove data-facets-visible as it was not used anymore
Removed the update trigger in showFacets as it is only called from toggleFacet wich already triggers the update

InnaBoitsun commented on src/Orckestra.Composer.Website/UI.Package/Typescript/Composer.ContentSearch/SearchResultsController.ts:
```diff
@@ -33,21 +39,72 @@ module Orckestra.Composer {
+                            this.showFacet();
+                        }
+                        FacetsVisible = !FacetsVisible;
+                        self.vueSearchResults.$data.FacetsVisible = FacetsVisible; // setting this will trigger the "updated" function above
```
do you need both variables - global and inside components?
also inside component, you can assign values like this
this.FacetsVisible = !this.FacetsVisible  
instead of
self.vueSearchResults.$data.FacetsVisible = ...

a-rousseau replied:
We need local and global variable because the local method "updated" will not be triggered if a local variable is not changed also it is used in SearchHelpers.cshtml

Going to try and removing the global one

Commit:
Removed the global variable FacetsVisible

Remove the obsolete global variable FacetsVisible

InnaBoitsun commented on src/Orckestra.Composer.Website/App_Data/Razor/Composer/Search/BaseSearchResults.cshtml:
```diff
@@ -24,6 +24,7 @@
 @{
     string vueId = "vueSearchResults" + Guid.NewGuid().ToString("N");
     string colNum = "";
+    bool isOverride = ProductRowNum != 3;
```
what is the purpose of this? please name it more informative

Commit:
Update BaseSearchResults.cshtml

rename variable

InnaBoitsun commented on src/Orckestra.Composer.Website/UI.Package/Typescript/Composer.ContentSearch/SearchResultsController.ts:
```diff
@@ -33,21 +38,70 @@ module Orckestra.Composer {
+                        self.showFacetsService.setShowFacets(this.FacetsVisible);
+                    },
+                    updateProductColumns(){
+                        let searchContainer = document.getElementsByClassName("search-container");
```
if we need some identifier for JS it is better to use id attribute for this, as class can be deleted
when you use id, you don'tt need to do this for loop
` for (let i=0; i < searchContainer.length; i++) {`
 ...
 
 also is it next correct? Do nothing when isOverriden is true?
 Please rename here also this variable what it means, to understand better
` if (isOverriden) return;`

a-rousseau replied:
It cannot be an ID because it is applied to the product tile container (grid columns) and we don't always know the number of tiles there will be.

![Screen Shot 2022-11-29 at 7 14 55 AM](https://user-images.githubusercontent.com/44198946/204527484-be244a73-a687-4baa-8e75-22834fbb2bd7.png)

InnaBoitsun replied:
Ok, then name is confusing - was thinking that this is a class for all search container, but this is just for tile
probably rename - product-col-container ?

InnaBoitsun commented on src/Orckestra.Composer.Website/UI.Package/Typescript/Composer.Product/ProductDetail/ProductDetailController.ts:
```diff
@@ -316,6 +318,71 @@ module Orckestra.Composer {
+                    }
+                },
+                methods: {
+                    HasImage(value:any) {
```
is it used? Does value has Imageurl ?

a-rousseau replied:
value has ImageUrl but it is not used, removing

InnaBoitsun commented on src/Orckestra.Composer.Website/UI.Package/Typescript/Composer.Product/ProductDetail/ProductDetailController.ts:
```diff
@@ -316,6 +318,71 @@ module Orckestra.Composer {
+                    Variant: self.context.viewModel.selectedVariantId,
+                    Values: self.context.viewModel.keyVariantAttributeItems,
+                    Loading: false,
+                    dataUpdatedTracker: 1
```
is this used?

Commit:
fix based on PR Comments

Removed unused variables in ProductDetailController
Removed isOverriden and use new logic to detect if resize should be done
Renamed search-container (searchContainer) to product-col-container (productColContainer)
fixed some minor css issues

Commit:
Moved popover trigger to inner div

Moved popover trigger to inner div to fix issues when hover was triggered twice on click and left the popover opened

Commit:
remove unnecessary variable

remove unnecessary Data variable in ProductDetailController

Commit:
Fix logic for specifications

Fixed cases when Groups might not be null but still empty
